### DISCUSSION TITLE
Public-key hash based mapping of policies in STS AssumeRoleWithCertificate

### DISCUSSION THREAD

**Discussion by nicopal** [minio/minio#17243] [Q&A] [answered]
Hi,

currently STS AssumeRoleWithCertificate supports mapping policies to Subject CommonName (implemented in sts-handlers.go). However, the current implementation does not add any security benefit, since  AssumeRoleWithCertificate authenticates any well-formed certificate (even self-signed) and the Subject CommonName can be easily spoofed. As a result, anyone who knows the policy name can use that as Subject Common name to create a self-signed X.509 certificate and obtain access.

However, mapping policies to the certificate public-key hash would bring a significant security capability improvements. Another very helpful feature would be **mapping policies to the issuer public-key hash**. In enterprise scenarios, that would help provide key identity-based policies for all certificates signed by e.g. an internal CA. The improvements mentioned above would allow fine-grained policies that apply to a specific **key identity** (certificate.PublicKey) or **issuer identity** (certificate.AuthorityKeyId).

Are there plans to implement public-key hash based mapping in STS AssumeRoleWithCertificate in the near future?

Alternatively, what is the quickest way to introduce this capability - a pull request with a patch? 

Thanks.

**Comment by harshavardhana** [comment]
@aead ^^

**Reply by ravindk89** [reply]
> what is the quickest way to introduce this capability - a pull request with a patch?

If you want to take a shot at it, we would welcome your contribution. Check out our [contributors guide](https://github.com/minio/minio/blob/master/CONTRIBUTING.md)

**Comment by aead** [comment, marked as answer]
@nicopal 
> currently STS AssumeRoleWithCertificate supports mapping policies to Subject CommonName (implemented in sts-handlers.go). However, the current implementation does not add any security benefit, since AssumeRoleWithCertificate authenticates any well-formed certificate (even self-signed) and the Subject CommonName can be easily spoofed. As a result, anyone who knows the policy name can use that as Subject Common name to create a self-signed X.509 certificate and obtain access.

Are you sure this is correct. If you take a look at (`sts-handlers.go`):
```
	if !globalIAMSys.STSTLSConfig.InsecureSkipVerify { // Verify whether the client certificate has been issued by a trusted CA.
		_, err := certificate.Verify(x509.VerifyOptions{
			KeyUsages: []x509.ExtKeyUsage{
				x509.ExtKeyUsageClientAuth,
			},
			Roots: globalRootCAs,
		})
		if err != nil {
			writeSTSErrorResponse(ctx, w, ErrSTSInvalidClientCertificate, err)
			return
		}
```
If you don't disable certificate verification in the STS AssumeRoleWithCertificate config, the certificate should be verified.

> However, mapping policies to the certificate public-key hash would bring a significant security capability improvements. Another very helpful feature would be mapping policies to the issuer public-key hash. In enterprise scenarios, that would help provide key identity-based policies for all certificates signed by e.g. an internal CA. The improvements mentioned above would allow fine-grained policies that apply to a specific key identity (certificate.PublicKey) or issuer identity (certificate.AuthorityKeyId).

> Are there plans to implement public-key hash based mapping in STS AssumeRoleWithCertificate in the near future?

In general, I agree that authentication based on certificate public key hashes is a reasonable authentication mechanism that works across platforms and without any special SDK functionality. The AssumeRoleWithCertificate STS authentication using certificate public key hashes just hasn't come up as requested feature so far. However, our [minio/kes](https://github.com/minio/kes) project uses this technique extensively.
I will take a look at bringing this also to MinIO 👍

**Reply by nicopal** [reply]
Thanks for supporting the idea! 🙌 also thanks for the pointer to KES and btw MinIO is awesome 👍
Adding more capabilities to certificate-based authentication would really help use MinIO in use cases that need advanced security. You're right that with the STS AssumeRoleWithCertificate config the certificate is verified. However, it's quite limited and coarse-grained, because it then maps certificates signed by a trusted CA to whatever policy is mentioned in the Subject CommonName (I was wrong about the comment on self-signed certificates in my original comment, sorry for that!). The problem here is that there can be several _trusted CA certificates_ and they might (or sometimes even should not) have the same level of trust. 

So some great-to-have features (in order of priority) would be:
1. mapping public key hashes/key id's **of the signing CA certificate** (certificate.AuthorityKeyId) to a specific policy. That would enable certificate-based group authentication - very useful (I think this is what [this](https://github.com/minio/minio/issues/17096) issue is about).
2. mapping public key hash/key id of a specific X509 certificate to a specific policy. This is fine-grained certificate-based authentication akin to attaching a policy to a certificate in amazon IoT core https://docs.aws.amazon.com/iot/latest/developerguide/attach-to-cert.html
